libvast/src/system/local_segment_store.cpp

```diff
@@ -0,0 +1,306 @@
+        ->request(self->state.fs, caf::infinite, atom::write_v, new_path,
+                  new_segment->chunk())
+        .then(
+          [seg = std::move(*new_segment), self, old_path,
+           new_path](atom::ok) mutable {
+            std::error_code ec;
+            // Re-use the old filename so that we don't have to write a new
+            // partition flatbuffer with the changed store header as well.
+            std::filesystem::rename(new_path, old_path, ec);
+            if (ec)
+              VAST_ERROR("failed to erase old data {}", seg.id());
+            self->state.segment = std::move(seg);
+          },
+          [](caf::error& err) {
+            VAST_ERROR("failed to flush archive {}", to_string(err));
+          });
+      return atom::done_v;
+    },
+  };
+}
+
+store_builder_actor::behavior_type active_local_store(
+  store_builder_actor::stateful_pointer<active_store_state> self,
+  filesystem_actor fs, const std::filesystem::path& path) {
+  VAST_INFO("spawning active local store active"); // FIXME: INFO -> DEBUG
+  // TODO: The shutdown path is copied from the archive; align it
+  // with the fs actor.
+  self->state.builder
+    = std::make_unique<segment_builder>(defaults::system::max_segment_size);
+  self->set_exit_handler([self, path, fs](const caf::exit_msg&) {
+    VAST_INFO("exiting active store");
+    auto seg = self->state.builder->finish();
+    self->request(fs, caf::infinite, atom::write_v, path, seg.chunk())
+      .then([](atom::ok) { /* nop */ },
+            [self](caf::error& err) {
+              VAST_ERROR("failed to flush archive {}", to_string(err));
+            });
+    self->quit();
+  });
+
+  return {
+    // store
+    [self](vast::query query, const ids& ids) -> caf::result<atom::done> {
+      auto slices = self->state.builder->lookup(ids);
+      if (!slices)
+        return slices.error();
+      if (caf::holds_alternative<query::erase>(query.cmd)) {
+        return self->delegate(static_cast<store_actor>(self), atom::erase_v,
+                              ids);
+      }
+      return handle_lookup(self, query, ids, *slices);
+    },
+    [self](atom::erase, const ids& ids) -> caf::result<atom::done> {
+      auto seg = self->state.builder->finish();
+      auto id = seg.id();
+      auto slices = seg.erase(ids);
+      if (!slices)
+        return slices.error();
+      self->state.builder->reset(id);
+      for (auto&& slice : std::exchange(*slices, {}))
+        self->state.builder->add(std::move(slice));
+      return atom::done_v;
+    },
+    // store builder
+    [self](
+      caf::stream<table_slice> in) -> caf::inbound_stream_slot<table_slice> {
+      return self
+        ->make_sink(
+          in, [=](caf::unit_t&) {},
+          [=](caf::unit_t&, std::vector<table_slice>& batch) {
+            VAST_TRACE("{} gets batch of {} table slices", self, batch.size());
+            for (auto& slice : batch)
+              if (auto error = self->state.builder->add(slice))
```

Comment:
Don't use a builder immediately here, but just add all table slices at once just before calling `finish()`.

Reply:
We currently rely on the builder to be able to call `lookup()` on it.